[Bug] SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x860100

### Is this a support request?

- [x] This is not a support request

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Current Behavior

After upgrading to 0.24.0 I get panics in headscale like such:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x860100]
goroutine 18595 [running]:
github.com/juanfont/headscale/hscontrol/types.(*Node).Proto(0x0)
	/home/runner/work/headscale/headscale/hscontrol/types/node.go:239 +0x20
github.com/juanfont/headscale/hscontrol/types.Routes.Proto({0x400077bc08?, 0x0?, 0x1cf16f0?})
	/home/runner/work/headscale/headscale/hscontrol/types/routes.go:82 +0xb0
github.com/juanfont/headscale/hscontrol.headscaleV1APIServer.GetRoutes({{}, 0x0?}, {0x19f2180?, 0x1?}, 0x40014101c0?)
	/home/runner/work/headscale/headscale/hscontrol/grpcv1.go:558 +0x3c
github.com/juanfont/headscale/gen/go/headscale/v1._HeadscaleService_GetRoutes_Handler({0x1b9f9e0, 0x4000718328}, {0x1fa5c58, 0x4000c2e3f0}, 0x4001104480, 0x0)
	/home/runner/work/headscale/headscale/gen/go/headscale/v1/headscale_grpc.pb.go:806 +0x1c0
google.golang.org/grpc.(*Server).processUnaryRPC(0x4000680600, {0x1fa5c58, 0x4000c2e390}, 0x4000782420, 0x400070f3b0, 0x30b7678, 0x0)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.69.0/server.go:1392 +0xc44
google.golang.org/grpc.(*Server).handleStream(0x4000680600, {0x1fa6418, 0x40001f41a0}, 0x4000782420)
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.69.0/server.go:1802 +0x910
google.golang.org/grpc.(*Server).serveStreams.func2.1()
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.69.0/server.go:1030 +0x84
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 102
	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.69.0/server.go:1041 +0x13c
```

The tailnet is working, in that sense that I can connect to all nodes as far as i can see. However, on the host system "docker exec headscale headscale nodes list" shows almost all nodes offline. Then, often I cant even use this command because it says the container is restarting. However, the tailnet keeps working. I just pinged a node on the tailnet that seemed offline, but after a single timeout the ping worked and another "docker exec headscale headscale nodes list" showed like half the nodes online. By this time the tailnet was up for minutes.

The this time I tried to get the nodes list I get:

`Error response from daemon: Container f56575937eba4f24b50471c1ea0521bc7fecda5705766dc40d11be2c983f048b is restarting, wait until the container is running`

Tailnet still working.

Since the golang error message mentions routes, could it be related to #90 ?

How would I check that?

### Expected Behavior

No panics. Having a reliable nodes list output. Being able to exec commands in the container.

### Steps To Reproduce

Start headscale docker container and check logs.

### Environment

```markdown
- OS: Linux raspberrypi 6.1.21-v8+ #1642 SMP PREEMPT Mon Apr  3 17:24:16 BST 2023 aarch64 GNU/Linux
- Headscale version: 0.24.0
- Tailscale version: many different versions on mac, iphone, windows and linux machines
```

### Runtime environment

- [x] Headscale is behind a (reverse) proxy
- [x] Headscale runs in a container

### Anything else?

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,6 +13,8 @@
 
 - Relax username validation to allow emails
   [#2364](https://github.com/juanfont/headscale/pull/2364)
+- Remove invalid routes and add stronger constraints for routes to avoid API panic
+  [#2371](https://github.com/juanfont/headscale/pull/2371)
 
 ## 0.24.0 (2025-01-17)
 

--- a/cmd/headscale/cli/routes.go
+++ b/cmd/headscale/cli/routes.go
@@ -251,10 +251,15 @@ func routesToPtables(routes []*v1.Route) pterm.TableData {
 			isPrimaryStr = strconv.FormatBool(route.GetIsPrimary())
 		}
 
+		var nodeName string
+		if route.GetNode() != nil {
+			nodeName = route.GetNode().GetGivenName()
+		}
+
 		tableData = append(tableData,
 			[]string{
 				strconv.FormatUint(route.GetId(), Base10),
-				route.GetNode().GetGivenName(),
+				nodeName,
 				route.GetPrefix(),
 				strconv.FormatBool(route.GetAdvertised()),
 				strconv.FormatBool(route.GetEnabled()),

--- a/hscontrol/db/db.go
+++ b/hscontrol/db/db.go
@@ -521,6 +521,27 @@ func NewHeadscaleDatabase(
 				},
 				Rollback: func(db *gorm.DB) error { return nil },
 			},
+			{
+				// Add a constraint to routes ensuring they cannot exist without a node.
+				ID: "202501221827",
+				Migrate: func(tx *gorm.DB) error {
+					// Remove any invalid routes associated with a node that does not exist.
+					if tx.Migrator().HasTable(&types.Route{}) && tx.Migrator().HasTable(&types.Node{}) {
+						err := tx.Exec("delete from routes where node_id not in (select id from nodes)").Error
+						if err != nil {
+							return err
+						}
+					}
+
+					err := tx.AutoMigrate(&types.Route{})
+					if err != nil {
+						return err
+					}
+
+					return nil
+				},
+				Rollback: func(db *gorm.DB) error { return nil },
+			},
 		},
 	)
 

--- a/hscontrol/types/routes.go
+++ b/hscontrol/types/routes.go
@@ -79,13 +79,16 @@ func (rs Routes) Proto() []*v1.Route {
 	for _, route := range rs {
 		protoRoute := v1.Route{
 			Id:         uint64(route.ID),
-			Node:       route.Node.Proto(),
 			Prefix:     route.Prefix.String(),
 			Advertised: route.Advertised,
 			Enabled:    route.Enabled,
 			IsPrimary:  route.IsPrimary,
 			CreatedAt:  timestamppb.New(route.CreatedAt),
 			UpdatedAt:  timestamppb.New(route.UpdatedAt),
+		}
+
+		if route.Node != nil {
+			protoRoute.Node = route.Node.Proto()
 		}
 
 		if route.DeletedAt.Valid {

--- a/hscontrol/types/routes.go
+++ b/hscontrol/types/routes.go
@@ -13,7 +13,7 @@ import (
 type Route struct {
 	gorm.Model
 
-	NodeID uint64
+	NodeID uint64 `gorm:"not null"`
 	Node   *Node
 
 	// TODO(kradalby): change this custom type to netip.Prefix

